mio/src/event_loop.rs

```diff
@@ -363,43 +364,47 @@ const ADD_SOURCE_TOKEN: usize = 0;
 pub struct AddSource {
     loop_handle: LoopHandle,
     source: Option<Source>,
-    id: Arc<AtomicUsize>,
-    scheduled: bool,
+    result: Option<(Arc<Slot<io::Result<usize>>>, slot::Token)>,
 }
 
 impl Future for AddSource {
     type Item = usize;
-    type Error = io::Error; // TODO: integrate channel error?
+    type Error = io::Error;
 
     fn poll(&mut self, tokens: &Tokens) -> Option<Result<usize, io::Error>> {
-        if self.scheduled {
-            if tokens.may_contain(&Tokens::from_usize(ADD_SOURCE_TOKEN)) {
-                let id = self.id.load(Ordering::Relaxed);
-                if id != 0 {
-                    return Some(Ok(id))
+        match self.result {
+            Some((ref result, _)) => {
+                if tokens.may_contain(&Tokens::from_usize(ADD_SOURCE_TOKEN)) {
+                    result.try_consume().ok()
+                } else {
+                    None
                 }
             }
-        } else {
-            if CURRENT_LOOP.is_set() {
-                let res = CURRENT_LOOP.with(|lp| {
-                    if lp.id == self.loop_handle.id {
-                        Some(lp.add_source(self.source.take().unwrap()))
-                    } else {
-                        None
-                    }
-                });
-                if let Some(id) = res {
-                    return Some(Ok(id));
-                }
+            None => {
+                let source = &mut self.source;
+                self.loop_handle.with_loop(|lp| {
+                    lp.map(|lp| {
+                        lp.add_source(source.take().unwrap())
+                    })
+                })
             }
         }
-
-        None
     }
 
     fn schedule(&mut self, wake: Arc<Wake>) {
-        if self.scheduled { return; }
-        self.scheduled = true;
-        self.loop_handle.add_source_(self.source.take().unwrap(), self.id.clone(), wake);
+        if let Some((ref result, ref mut token)) = self.result {
+            result.cancel(*token);
```

Comment:
What about the case where the previous callback is invoked just before we cancel it? The wake we're about to register will then never be invoked, meaning we may never wake up to realize the id is ready.